### PR TITLE
WIP:  Add pre-fetching to RowServiceInputStream

### PR DESCRIPTION
**Pull request by jpmcmu**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [ ] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana WIP pre-fetching code

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 1]
```diff
@@ -344,12 +329,20 @@ private Object parseFlatField(FieldDef fd, boolean isLittleEndian) throws Unpars
                     dataLen = (int) getInt(4, isLittleEndian);
                 }
 
-                if (dataLen > this.inputStream.available())
+                byte[] bytes = new byte[dataLen];
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java]: why do the other field types read into this.scratchBuffer ?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 2]
```diff
@@ -344,12 +329,20 @@ private Object parseFlatField(FieldDef fd, boolean isLittleEndian) throws Unpars
...
+
+                int bytesConsumed = 0;
+                while (bytesConsumed < dataLen)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java]: in what type of scenario do we end up iterating more than once here?

**Comment by rpastrana** [comment]
@jpmcmu associate this PR with https://track.hpccsystems.com/browse/JAPI-240

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 3]
```diff
@@ -360,29 +450,39 @@ void readBlock() throws HpccFileException
...
-            // Compact the array before reading in new data
-            int remainingBytesInBuffer = this.readBufferLen - this.readPos;
-            this.readBufferLen = remainingBytesInBuffer + dataLen;
+            while (dataLen > 0)
+            {
+                try
+                {
+                    this.bufferWriteMutex.acquire();
+                }
+                catch(Exception e){}
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: does this catch an interrupt?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 4]
```diff
@@ -312,22 +366,42 @@ void readBlock() throws HpccFileException
...
 
         try
         {
             this.handle = dis.readInt();
             if (this.handle == 0)
             {
-                len = retryWithCursor();
+                try
+                {
+                    len = retryWithCursor();
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: I can't tell if retryWithCursor generates a trace log entry, but if it doesn't let's make sure we notify we're trying to handle a restart due to the received 0 handle

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 5]
```diff
@@ -360,29 +450,39 @@ void readBlock() throws HpccFileException
...
             }
 
-            // Compact the array before reading in new data
-            int remainingBytesInBuffer = this.readBufferLen - this.readPos;
-            this.readBufferLen = remainingBytesInBuffer + dataLen;
+            while (dataLen > 0)
+            {
+                try
+                {
+                    this.bufferWriteMutex.acquire();
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: do we know for sure we have enough room in the buffer at this point? Could we land here after sleeping (line 481)?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 6]
```diff
@@ -360,29 +450,39 @@ void readBlock() throws HpccFileException
...
-            if (dataLen > this.readBuffer.length) {
-                throw new HpccFileException("Error during read block, available data len is larger than read buffer. This should not happen");
+                // ReadBufferLen should only ever be updated within the mutex so no need to do compare exchange
+                this.readBufferLen.set(currentBufferLen + bytesToRead);
+                dataLen -= bytesToRead;
+
+                bufferWriteMutex.release();
+
+                // If we don't have room in the buffer to fetch more data sleep
+                if (dataLen > 0  && this.readBufferLen.get() > (BUFFER_PREFETCH_THRESHOLD_KB * 1024))
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: do we need to ensure another thread hasn't written to the buffer, but hasn't set the readbufferlen yet (line 471)?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 7]
```diff
@@ -413,22 +518,80 @@ void readBlock() throws HpccFileException
...
+            {
+                Thread.sleep(READ_WAIT_SLEEP_MS);
+            }
+        }
+        catch(Exception e){}
+    }
+
+    public void compactBuffer()
+    {
+        if (this.readPos >= BUFFER_COMPACT_THRESHOLD_KB * 1024)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: it feels like read and write access to this.readPos, this.markPos, and others should be protected

**Comment by jpmcmu** [comment]
@rpastrana closing this PR. New PR created under correct issue #: https://github.com/hpcc-systems/HPCC-JAPIs/pull/338

**Pull request closed without merging**